jetbrains-core/src/software/aws/toolkits/jetbrains/core/stack/Updater.kt

```diff
@@ -0,0 +1,111 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package software.aws.toolkits.jetbrains.core.stack
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.Application
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.project.Project
+import com.intellij.util.Alarm
+import com.intellij.util.AlarmFactory
+import software.amazon.awssdk.core.exception.SdkException
+import software.amazon.awssdk.services.cloudformation.CloudFormationClient
+import software.amazon.awssdk.services.cloudformation.model.DescribeStackResourcesRequest
+import software.amazon.awssdk.services.cloudformation.model.DescribeStacksRequest
+import software.amazon.awssdk.services.cloudformation.model.StackEvent
+import software.amazon.awssdk.services.cloudformation.model.StackResource
+import software.amazon.awssdk.services.cloudformation.model.StackStatus
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import javax.swing.SwingUtilities
+
+/**
+ * CallBacks for Updater. Called on EDT.
+ */
+internal interface UpdateListener {
+    fun onError(message: String)
+    fun onStackStatusChanged(stackStatus: StackStatus)
+}
+
+/**
+ * Updates tree view on timer. Create on EDT.
+ */
+internal class Updater(
+    private val treeView: TreeView,
+    private val project: Project,
+    private val eventsTableView: TableView,
+    private val stackName: String,
+    private val updateEveryMs: Int,
+    private val listener: UpdateListener
+) : Disposable {
+
+    @Volatile
+    private var previousStackStatusType: StatusType = StatusType.UNKNOWN
+    // Access from Pool only
+    private val eventsFetcher = EventsFetcher(stackName)
+
+    private val client get() = AwsClientManager.getInstance(project).getClient<CloudFormationClient>()
```

Comment:
You should cache this on the tool window's content creation and re-use it. That way if they switch the active region the tool window doesn't glitch out

Reply:
Ok. I cached it on ``Updater`` level, so one ``Updater`` -- one stack